source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,192 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+    return std::move(handle);
+  }
+  ENVOY_LOG(
+      debug,
+      "cm odcds: forwarding the on-demand discovery request for cluster {} to the main thread",
+      name);
+  // This seems to be the first request for discovery of this cluster in this worker thread. Rest of
+  // the process may only happen in the main thread.
+  dispatcher_.post([this, odcds = std::move(odcds), timeout,
+                    weak_callback = ClusterDiscoveryCallbackWeakPtr(callback), name,
+                    &thread_local_dispatcher = cluster_manager.thread_local_dispatcher_] {
+    // Check for the cluster here too. It might have been added between the time when this closure
+    // was posted and when it is being executed.
+    if (getThreadLocalCluster(name) != nullptr) {
+      ENVOY_LOG(debug, "cm odcds: the requested cluster {} is already known", name);
+      if (weak_callback.expired()) {
+        ENVOY_LOG(debug, "cm odcds: callback expired");
+        // Not only the cluster was added, but it was also already handled, so don't bother with
+        // posting the callback back to the worker thread.
+        return;
+      }
+      thread_local_dispatcher.post([name, weak_callback] {
+        if (auto callback = weak_callback.lock(); callback != nullptr) {
+          ENVOY_LOG(debug, "cm odcds: invoking callback on the already known cluster {}", name);
+          // If this gets called here, it means that we requested a discovery of a cluster without
+          // checking if that cluster is already known by cluster manager. The other situation we
+          // get here can be when another thread was faster in requesting the discovery and handling
+          // the resulting new cluster. It's a rather unlikely race, though.
+          (*callback)(ClusterDiscoveryStatus::Available);
+        }
+      });
+      return;
+    }
+
+    if (auto it = pending_cluster_creations_.find(name); it != pending_cluster_creations_.end()) {
+      ENVOY_LOG(debug, "cm odcds: on-demand discovery for cluster {} is already in progress", name);
+      // We already began the discovery process for this cluster, nothing to do. If we got here, it
+      // means that it was other worker thread that requested the discovery.
+      return;
+    }
+    // Start the discovery. If the cluster gets discovered, cluster manager will warm it up and
+    // invoke the cluster lifecycle callbacks, that will in turn invoke our callback.
+    odcds->updateOnDemand(name);
+    // Setup the discovery timeout timer to avoid keeping callbacks indefinitely.
+    auto timer = dispatcher_.createTimer([this, name] { notifyExpiredDiscovery(name); });
+    timer->enableTimer(timeout);
+    // Keep odcds handle alive for the duration of the discovery process.
+    pending_cluster_creations_.insert(
+        {std::move(name), ClusterCreation{std::move(odcds), std::move(timer)}});
+  });
+
+  return std::move(handle);
+}
+
+void ClusterManagerImpl::notifyExpiredDiscovery(const std::string& name) {
```

Comment:
debug log please

Reply:
I have added there a debug log already - `ENVOY_LOG(debug, "cm odcds: on-demand discovery for cluster {} timed out", name);`. Did github place this comment in wrong place?